pkg/controller/smart-clone-controller.go

```diff
@@ -1,377 +1,234 @@
-			return err
-		}
+		return reconcile.Result{}, err
 	}
-	return nil
+	return r.reconcilePvc(log, pvc)
 }
 
-func (c *SmartCloneController) syncSnapshot(key string) error {
-	snapshot, exists, err := c.snapshotFromKey(key)
-	if err != nil {
-		return err
-	} else if !exists {
-		return nil
-	}
+func (r *SmartCloneReconciler) reconcilePvc(log logr.Logger, pvc *corev1.PersistentVolumeClaim) (reconcile.Result, error) {
+	log.WithValues("pvc.Name", pvc.Name).WithValues("pvc.Namespace", pvc.Namespace).Info("PVC created from snapshot, updating datavolume status")
+	snapshotName := pvc.Spec.DataSource.Name
 
-	_, ok := snapshot.Annotations[AnnSmartCloneRequest]
-	if !ok {
-		//ignoring snapshot, not created by DataVolume Controller
-		return nil
-	}
-
-	snapshotReadyToUse := snapshot.Status.ReadyToUse
-	klog.V(3).Infof("Snapshot \"%s/%s\" - ReadyToUse: %t", snapshot.Namespace, snapshot.Name, snapshotReadyToUse)
-	if !snapshotReadyToUse {
-		return nil
-	}
-
-	return c.processNextSnapshotItem(snapshot)
-}
-
-// The snapshot is ReadyToUse, then we can create the PVC and update DV status
-func (c *SmartCloneController) processNextSnapshotItem(snapshot *csisnapshotv1.VolumeSnapshot) error {
-	dataVolume, err := c.dataVolumeLister.DataVolumes(snapshot.Namespace).Get(snapshot.Name)
-	if err != nil {
-		return err
+	datavolume := &cdiv1.DataVolume{}
+	if err := r.Client.Get(context.TODO(), types.NamespacedName{Name: snapshotName, Namespace: pvc.Namespace}, datavolume); err != nil {
+		return reconcile.Result{}, err
 	}
 
 	// Update DV phase and emit PVC in progress event
-	err = c.updateSmartCloneStatusPhase(SmartClonePVCInProgress, dataVolume, nil)
-	if err != nil {
+	if err := r.updateSmartCloneStatusPhase(cdiv1.Succeeded, datavolume, pvc); err != nil {
 		// Have not properly updated the data volume status, don't delete the snapshot so we retry.
-		klog.Errorf("error updating datavolume with success, requeuing: %v", err)
-		return err
+		log.Error(err, "error updating datavolume with success")
+		return reconcile.Result{}, err
 	}
 
-	newPvc := newPvcFromSnapshot(snapshot, dataVolume)
-	if newPvc == nil {
-		klog.Errorf("error creating new pvc from snapshot object")
-		return nil
+	snapshotToDelete := &csiv1.VolumeSnapshot{}
+	if err := r.Client.Get(context.TODO(), types.NamespacedName{Name: snapshotName, Namespace: pvc.Namespace}, snapshotToDelete); err != nil {
+		if k8serrors.IsNotFound(err) {
+			// Already gone, so no need to try a delete.
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
 	}
 
-	klog.V(3).Infof("Creating PVC \"%s/%s\" from snapshot", newPvc.Namespace, newPvc.Name)
-	_, err = c.clientset.CoreV1().PersistentVolumeClaims(snapshot.Namespace).Create(newPvc)
-	if err != nil {
-		klog.Errorf("error creating pvc from snapshot: %v", err)
-		return err
+	if err := r.Client.Delete(context.TODO(), snapshotToDelete); err != nil {
+		log.Error(err, "error deleting snapshot for smart-clone")
```

Comment:
why not handle notfound here as well

Reply:
I check for not found right above, but you are right it is possible that in between those two statements the snapshot got deleted, so I should not error if the delete fails on a not found.